tests/vmi_vsock_test.go

```diff
@@ -0,0 +1,96 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2022 Red Hat, Inc.
+ *
+ */
+
+package tests_test
+
+import (
+	"encoding/xml"
+
+	expect "github.com/google/goexpect"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"k8s.io/utils/pointer"
+
+	virtconfig "kubevirt.io/kubevirt/pkg/virt-config"
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/api"
+	"kubevirt.io/kubevirt/tests/framework/checks"
+
+	"kubevirt.io/client-go/kubecli"
+
+	"kubevirt.io/kubevirt/tests"
+	"kubevirt.io/kubevirt/tests/console"
+)
+
+var _ = Describe("[Serial][sig-compute]Vsock", func() {
+	var virtClient kubecli.KubevirtClient
+	var err error
+
+	BeforeEach(func() {
+		virtClient, err = kubecli.GetKubevirtClient()
+		Expect(err).ToNot(HaveOccurred())
+	})
+
+	Context("", func() {
+		It("should expose a Vsock device", func() {
+			checks.SkipTestIfNoFeatureGate(virtconfig.VsockGate)
+			By("Creating a VMI with Vsock enabled")
+			vmi := tests.NewRandomFedoraVMI()
+			vmi.Spec.Domain.Devices.AutoattachVsock = pointer.Bool(true)
+			vmi = tests.RunVMIAndExpectLaunch(vmi, 60)
+
+			By("creating valid libvirt domain")
+
+			domain, err := tests.GetRunningVirtualMachineInstanceDomainXML(virtClient, vmi)
+			Expect(err).ToNot(HaveOccurred())
+			domSpec := &api.DomainSpec{}
+			Expect(xml.Unmarshal([]byte(domain), domSpec)).To(Succeed())
+			Expect(domSpec.Devices.Vsock.CID.Auto).To(Equal("no"))
+			Expect(domSpec.Devices.Vsock.CID.Address).To(Equal("3"))
+
+			By("Logging in as root")
+			err = console.LoginToFedora(vmi)
+			Expect(err).ToNot(HaveOccurred())
+
+			By("Ensuring a vsock device is present")
+			Expect(console.SafeExpectBatch(vmi, []expect.Batcher{
+				&expect.BSnd{S: "ls /dev/vsock-vhost\n"},
+				&expect.BExp{R: "/dev/vsock-vhost"},
+			}, 300)).To(Succeed(), "Could not find a vsock-vhost device")
+			Expect(console.SafeExpectBatch(vmi, []expect.Batcher{
+				&expect.BSnd{S: "ls /dev/vsock\n"},
+				&expect.BExp{R: "/dev/vsock"},
+			}, 300)).To(Succeed(), "Could not find a vsock device")
+
+			By("Migrating the VMI")
+			checks.SkipIfMigrationIsNotPossible()
+			migration := tests.NewRandomMigration(vmi.Name, vmi.Namespace)
+			tests.RunMigrationAndExpectCompletion(virtClient, migration, tests.MigrationWaitTime)
+
+			By("Ensuring the Vsock is still present")
+			Expect(console.SafeExpectBatch(vmi, []expect.Batcher{
```

Comment:
After considering the comment from @vladikr I am not sure if libvirt would pick up the new vosck CID on the new server. There is a good chance that it kept the one from the old CID (which in this case accidentally matches the new assigned CID).

Reply:
From my understanding after reading [vsock man page](https://man7.org/linux/man-pages/man7/vsock.7.html) if the CID is unavailable on the host after `live-migration` it will be assigned with a new CID and it will notify the guest about it.

Reply:
I think this only means that live migration is supported and the hypervisor may have to give it a different cid. It does not mean that you get from the kernel another cid without involving the hypervisor (at least pretty sure).

Reply:
I added 2 test cases to verify the live migration behavior. However, the behavior is weird.
I ran the tests on a 2 nodes setup.

1. Create 2 VMs on node02. The VMs will have CID 3, and 4. Migrate the the VM with CID 4. If libvirt retain the CID then the migrated VM should still have CID 4. However, it has CID 3.
2. Create a VM on node02. Hack the cache file on node01 to reserve CID 3 from virt-handler's allocation. Migrate the VM. Now virt-handler allocation CID 4 to the migration target but the libvirt domain reports CID 3.

I seems that libvirt doesn't retain the CID for migration target but just assign an available CID to the migration target regardless of the domain spec.

Reply:
Sorry my previous comment is not correct. Libvirt does retain the CID if available, which is demonstrated in the test case.

Also the email thread confirms the behavior https://www.spinics.net/lists/netdev/msg631998.html.

Reply:
> Also the email thread confirms the behavior https://www.spinics.net/lists/netdev/msg631998.html.

Pefect. That is how it should be. What I did not think about is, that the source libvirt is defining the destionation domain xml during migration. This is a little bit cumbersome for us, since we have to transfer additional info (like here the new CID) back to the source handler, so that we can set it on the target ...

We do that for some info already and the final domain modification is happening at https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-launcher/virtwrap/live-migration-source.go#L257 (you can find it when following the `MigrateVMI` grcp call in virt-launcher.

I think our best option is to add the CID to the VMI status and let' virt-controller hand out the CIDs. We don't have a back-flow channel between target-handler and source-handler. It would be similar to the target CPU set.

Reply:
+1

Reply:
> I think our best option is to add the CID to the VMI status and let' virt-controller hand out the CIDs. We don't have a back-flow channel between target-handler and source-handler. It would be similar to the target CPU set.

I don't think it should be the `virt-controller` hand out the CIDs. The CIDs mapping should be persisted and `virt-controller` doesn't have the persistent directory. To follow the CPU set pattern, I think we can allocate the CID in [`migrationTargetUpdateVMIStatus`](https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-handler/vm.go#L729). There will be a `TargetCID` in `VirtualMachineInstanceMigrationState`. Then in https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-launcher/virtwrap/live-migration-source.go#L257, we can update the domXML with the new CID.

Reply:
It may just be much simpler for you to do it in virt-controller. The status would be your persistent storage and you have an initial sync "guarantee" after the cache sync where you can warm up your in-memory-view of the CIDs allocated. But what you propose will work too. Going with virt-controller just feels more robust and means no possible file-leftovers or filing cleanup steps. You also don't have to change the CID between miagrations and don't have to send that info back-and-forth.

Reply:
@PiotrProkop if I understand you correctly, you plan to connect from the VMs to the node, so you don't need the CID somewhere published for comparsion?

Reply:
> It may just be much simpler for you to do it in virt-controller. The status would be your persistent storage and you have an initial sync "guarantee" after the cache sync where you can warm up your in-memory-view of the CIDs allocated. But what you propose will work too. Going with virt-controller just feels more robust and means no possible file-leftovers or filing cleanup steps. You also don't have to change the CID between miagrations and don't have to send that info back-and-forth.

Thanks for the suggestion. This makes sense to me and I also want to simplify things. I will make the corresponding change.

Reply:
Updated to have virt-controller allocate the CIDs.

Reply:
No, it got never merged in the kernel. We had initially the same thought 💭